src/coreclr/src/jit/importer.cpp

```diff
@@ -9121,7 +9123,12 @@ GenTree* Compiler::impFixupStructReturnType(GenTree* op, CORINFO_CLASS_HANDLE re
         if ((op1->gtOper == GT_ADDR) && (op1->AsOp()->gtOp1->gtOper != GT_INDEX))
         {
             // Change '*(&X)' to 'X' and see if we can do better
-            op = op1->AsOp()->gtOp1;
+            bool canCSE = op->CanCSE();
+            op          = op1->AsOp()->gtOp1;
+            if (canCSE)
+            {
+                op->ClearDoNotCSE();
```

Comment:
It isn't necessary to guard this call to ClearDoNotCSE()

Reply:
Why is not it? It could be a volatile LCL_VAR that needs that flag.

Reply:
The implementation of CanCSE is 
` return ((gtFlags & GTF_DONT_CSE) == 0);`
So this is basically saying if the flag is set then clear it using the AND NOT operation.
Not sure how a volatile LCL_VAR would behave any differently here

Reply:
> So this is basically saying if the flag is set then clear it using the AND NOT operation.

Maybe I am missing something, but I don't understand.
We are checking the flag on `IND(ADDR(X))` tree and clear it on `X`. `X` was marked as `DONT_CSE` because it was `ADDR` source, `IND(ADDR(X))` could be marked as `DONT_CSE` for other reasons that we don't change here, so if `IND(ADDR(X))` was marked with that flag then after the transformation `X` should have this flag as well.

For example, `IND1(ADDR2(IND3(ADDR4(X))` in the beginning `X` and `IND3` have `DONT_CSE` because they are `ADDR` sources, when we optimize `IND3(ADDR4(X))` we have to keep `DONT_CSE` on `X` .

Reply:
Yes, I see.
I missed that we were overwriting `op` after setting `canCSE`.
This sequence is confusing and could use a comment, saying we are preserving any DONT_CSE that was set on the original version of `op`

```
 bool canCSE = op->CanCSE();
          op        = op1->AsOp()->gtOp1;
```